### ISSUE TITLE
AttributeError: 'str' object has no attribute 'user_id'

### ISSUE THREAD

**Issue by Matrixcoffee** [matrix-org/matrix-python-sdk#206]
As promised:

```
Last event received before exception:
{'content': {'avatar_url': None,
             'displayname': 'silentium (IRC)',
             'membership': 'join'},
 'event_id': '$152473819732408dTnHU:irc.hackint.org',
 'membership': 'join',
 'origin_server_ts': 1524738197452,
 'room_id': '!MbjuckLxNFDaNtXpWA:irc.hackint.org',
 'sender': '@irc_silentium:irc.hackint.org',
 'state_key': '@irc_silentium:irc.hackint.org',
 'type': 'm.room.member',
 'unsigned': {'age': 401,
              'prev_content': {'membership': 'leave'},
              'prev_sender': '@irc_silentium:irc.hackint.org',
              'replaces_state': '$152422114812799vOwiG:irc.hackint.org'}}
Stack trace:
Traceback (most recent call last):
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-python-sdk/matrix_client/client.py", line 425, in listen_forever
    self._sync(timeout_ms)
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-client-core/matrix_client_core/__init__.py", line 200, in _sync
    matrix_client.client.MatrixClient._sync(self, *args, **kwargs)
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-python-sdk/matrix_client/client.py", line 536, in _sync
    room._put_event(event)
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-python-sdk/matrix_client/room.py", line 310, in _put_event
    self._process_state_event(event)
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-python-sdk/matrix_client/room.py", line 693, in _process_state_event
    econtent.get("displayname"))
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-python-sdk/matrix_client/room.py", line 532, in _mkmembers
    if member.user_id not in [x.user_id for x in self._members]:
  File "/media/0E57-48B0/Matrix/Python-Matrix/matrix-python-sdk/matrix_client/room.py", line 532, in <listcomp>
    if member.user_id not in [x.user_id for x in self._members]:
AttributeError: 'str' object has no attribute 'user_id'
```

**Comment by Matrixcoffee** [comment]
Note that the event may or may not be accurate. matrix-client-core *tries* to capture the last event, but if the exception happens before the callback can grab the event, it'll be the previous, unrelated event.

**Comment by non-Jedi** [comment]
@Matrixcoffee Are you sure matrix-client-core isn't modifying the room objects `_members` attribute somewhere? That attribute used to be a dictionary mapping `user_id`s to `displayname`s, but now it's a simple list of `User` objects.

**Comment by non-Jedi** [comment]
If not, and you can trigger this deterministically, I'd love to see what `room._members` looks like.

**Comment by non-Jedi** [comment]
Likely fixed in #231
